blackjax/mcmc/barker.py

```diff
@@ -0,0 +1,306 @@
+from typing import Callable, NamedTuple
+
+import jax
+import jax.numpy as jnp
+from jax.scipy import stats
+from jax.tree_util import tree_flatten
+
+from blackjax.base import SamplingAlgorithm
+from blackjax.types import ArrayLikeTree, ArrayTree, PRNGKey
+
+__all__ = ["BarkerState", "BarkerInfo", "init", "build_kernel", "barker_proposal"]
+
+
+class BarkerState(NamedTuple):
+    """State of the Barker's proposal algorithm.
+
+    The Barker algorithm takes one position of the chain and returns another
+    position. In order to make computations more efficient, we also store
+    the current log-probability density as well as the current gradient of the
+    log-probability density.
+
+    """
+
+    position: ArrayTree
+    logdensity: float
+    logdensity_grad: ArrayTree
+
+
+class BarkerInfo(NamedTuple):
+    """Additional information on the Barker's proposal kernel transition.
+
+    This additional information can be used for debugging or computing
+    diagnostics.
+
+    proposal
+        The proposal that was sampled.
+    acceptance_rate
+        The acceptance rate of the transition.
+    is_accepted
+        Whether the proposed position was accepted or the original position
+        was returned.
+
+    """
+
+    acceptance_rate: float
+    is_accepted: bool
+    proposal: BarkerState
+
+
+def init(position: ArrayLikeTree, logdensity_fn: Callable) -> BarkerState:
+    grad_fn = jax.value_and_grad(logdensity_fn)
+    logdensity, logdensity_grad = grad_fn(position)
+    return BarkerState(position, logdensity, logdensity_grad)
+
+
+def build_kernel():
+    """Build a Barker's proposal kernel.
+
+    Returns
+    -------
+    A kernel that takes a rng_key and a Pytree that contains the current state
+    of the chain and that returns a new state of the chain along with
+    information about the transition.
+
+    """
+
+    def _compute_acceptance_probability(
+        logdensity: float,
+        logdensity_proposal: float,
+        logdensity_grad: ArrayTree,
+        logdensity_grad_proposal: ArrayTree,
+        position: ArrayTree,
+        position_proposal: ArrayTree,
+        scale: float,
+    ) -> float:
+        """Compute the acceptance probability of the Barker's proposal kernel."""
+
+        logdensity_grad, _ = tree_flatten(logdensity_grad)
+        logdensity_grad_proposal, _ = tree_flatten(logdensity_grad_proposal)
+        position, _ = tree_flatten(position)
+        position_proposal, _ = tree_flatten(position_proposal)
+
+        def ratio_proposal_nd(y, x, log_y, log_x):
+            num = -_log1pexp(-log_y * (x - y))
+            den = -_log1pexp(-log_x * (y - x))
+
+            return jnp.sum(num - den)
+
+        ratios_proposals = map(
+            ratio_proposal_nd,
+            position_proposal,
+            position,
+            logdensity_grad_proposal,
+            logdensity_grad,
+        )
+        ratio_proposal = sum(ratios_proposals)
```

Comment:
```suggestion
        ratios_proposals = jax.tree_util.tree_map(
            ratio_proposal_nd,
            position_proposal,
            position,
            logdensity_grad_proposal,
            logdensity_grad,
        )
        ratio_proposal = sum(jax.tree_util.tree_leaves(ratios_proposals))
```

Reply:
This is a list of arrays transformed into a list of floats. Is it worth the tree_map/tree_leaves?

Reply:
happy to change it, but then it felt a bit unnecessary as I have already flattened the tree, that's why I didn't do it.

Reply:
`position` could be a pytree right? tree_map/tree_leaves is usually pretty low cost

Reply:
Ah I see you have a bunch of tree_flatten in the begining - please remove those and use tree_map here then

Reply:
I don't know why I did that :D I guess it was late :)